source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,274 @@
+  return length >= MinimumContentLength ? length : MinimumContentLength;
+}
+
+uint64_t GzipFilterConfig::memoryLevelUint(Protobuf::uint32 level) {
+  return level > 0 ? level : DefaultMemoryLevel;
+}
+
+uint64_t GzipFilterConfig::windowBitsUint(Protobuf::uint32 window_bits) {
+  return (window_bits > 0 ? window_bits : DefaultWindowBits) | GzipHeaderValue;
+}
+
+GzipFilter::GzipFilter(const GzipFilterConfigSharedPtr& config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  skip_compression_ = !isAcceptEncodingAllowed(headers);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding() && isEtagAllowed(headers)) {
+
+    insertVaryHeader(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isAcceptEncodingAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* accept_encoding = headers.AcceptEncoding();
+  // TODO(gsagula): Since gzip is the only available content-encoding in Envoy at the moment,
+  // order/priority of preferred server encodings is disregarded (RFC2616-14.3). Replace this
+  // with a data structure that parses Accept-Encoding values and allows fast lookup of
+  // key/priority. Also, this should be part of some utility library.
+  if (accept_encoding) {
+    bool is_identity{true};  // true if does not exist or is followed by `q=0`.
+    bool is_wildcard{false}; // true if found and not followed by `q=0`.
+    bool is_gzip{false};     // true if exists and not followed by `q=0`.
+
+    for (const auto token :
+         StringUtil::splitToken(headers.AcceptEncoding()->value().c_str(), ",", true)) {
+      const auto values = StringUtil::splitToken(token, ";");
```

Comment:
PTAL at BM_FindTokenValueSearchForEqual for a better pattern, I think, for a similar use-case.  This just got merged this morning: https://github.com/envoyproxy/envoy/blob/590042324410597473f3e62edf275a5dc71f78f8/test/common/common/utility_speed_test.cc#L135

In the benchmark, compiled with-c opt, that pattern is faster than the nested split by ~3x in the benchmark.  See benchmark numbers in the PR description: https://github.com/envoyproxy/envoy/pull/2348

The key observation is that creating the extra level of vector is a little costly and we can just search for the delimiter and crop, rather than a general split.  Unless there can be more than one ; in which case I think the logic below may be wrong :)

Reply:
Benchmark looks awesome! I'll take a look a later.

Let me try to get rid of the second split call.

Reply:
Thanks!  I mean to ask for you to microoptimize in this PR, but maybe put a TODO here.  Right now I just want to make sure it's correct and we can flag any potential optimization points.